src/apdu_handler.c

```diff
@@ -0,0 +1,359 @@
+/*******************************************************************************
+*   (c) 2018, 2019 Zondax GmbH
+*   (c) 2016 Ledger
+*
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+********************************************************************************/
+
+#include <string.h>
+#include <os_io_seproxyhal.h>
+#include <os.h>
+
+#include "app_main.h"
+#include "cx.h"
+#include "crypto_helpers.h"
+#include "common.h"
+#include "addr.h"
+#include "ux.h"
+#include "actions.h"
+#include "coin.h"
+#include "view.h"
+#include "common/tx.h"
+#include "crypto.h"
+#include "zxmacros.h"
+#include "apdu_codes.h"
+#include "bech32.h"
+#include "app_mode.h"
+
+uint16_t action_addrResponseLen;
+
+__Z_INLINE uint8_t extractHRP(uint32_t rx, uint32_t offset) {
+    if (rx < offset + 1) {
+        THROW(APDU_CODE_DATA_INVALID);
+    }
+    MEMZERO(bech32_hrp, MAX_BECH32_HRP_LEN);
+
+    bech32_hrp_len = G_io_apdu_buffer[offset];
+
+    if (bech32_hrp_len == 0 || bech32_hrp_len > MAX_BECH32_HRP_LEN) {
+        THROW(APDU_CODE_DATA_INVALID);
+    }
+
+    memcpy(bech32_hrp, G_io_apdu_buffer + offset + 1, bech32_hrp_len);
+    bech32_hrp[bech32_hrp_len] = 0;     // zero terminate
+
+    return bech32_hrp_len;
+}
+
+__Z_INLINE void extractHDPath(uint32_t rx, uint32_t offset) {
+    if ((rx - offset) < sizeof(uint32_t) * HDPATH_LEN_DEFAULT) {
+        THROW(APDU_CODE_WRONG_LENGTH);
+    }
+
+    MEMCPY(hdPath, G_io_apdu_buffer + offset, sizeof(uint32_t) * HDPATH_LEN_DEFAULT);
+    
+    // Check values
+    if (hdPath[0] != HDPATH_0_DEFAULT ||
+        hdPath[1] != HDPATH_1_DEFAULT ||
+        hdPath[2] != HDPATH_2_DEFAULT) {
```

Comment:
I think I should remove check on `hdPath[2]` value has the app should allow multiple accounts.